pkg/lib/cockpit-components-select.jsx

```diff
@@ -20,125 +20,184 @@
+        constructor() {
+            super();
+
+            this.state = {
                 open: false,
-                currentData: this.props.initial,
                 documentClickHandler: undefined,
             };
-        },
-        clickHandler: function(ev) {
+        }
+
+        componentDidMount() {
+            const handler = this.handleDocumentClick.bind(this, React.findDOMNode(this));
+            this.setState({ documentClickHandler: handler });
+            document.addEventListener('click', handler, false);
+        }
+
+        componentWillUnmount() {
+            document.removeEventListener('click', this.state.documentClickHandler, false);
+        }
+
+        handleDocumentClick(node, ev) {
+            // clicking outside the select control should ensure it's closed
+            if (!node.contains(ev.target)) {
+                this.setState({ open: false });
+            }
+        }
+
+        clickHandler(ev) {
             // only consider clicks with the primary button
-            if (ev && ev.button !== 0)
+            if (ev && ev.button !== 0) {
                 return;
-            if (ev.target.tagName == 'A') {
-                var liElement = ev.target.offsetParent;
-                var elementData;
-                if ('data-data' in liElement.attributes)
+            }
+
+            if (ev.target.tagName === 'A') {
+                const liElement = ev.target.offsetParent;
+                let elementData;
+                if ('data-data' in liElement.attributes) {
                     elementData = liElement.attributes['data-data'].value;
+                }
 
                 this.setState({ open: false });
                 // if the item didn't change, don't do anything
-                if (elementData === this.state.currentData)
+                if (elementData === this.props.selected) {
                     return;
-                this.setState({ currentData: elementData });
-                if (this.props.onChange)
+                }
+                if (this.props.onChange) {
                     this.props.onChange(elementData);
+                }
             } else {
                 this.setState({ open: !this.state.open });
             }
-        },
-        render: function() {
-            var self = this;
-            var currentValue;
-
-            var listItems = React.Children.map(this.props.children, function(itm) {
-                var data = ('data' in itm.props) ? itm.props.data : undefined;
-                // we need to have some kind of value
-                var value = (itm.props.children !== undefined) ? itm.props.children : textForUndefined;
-                if (data === self.state.currentData)
-                    currentValue = value;
-                // if there's no initial value, use the first one
-                else if (!self.props.initial && currentValue === undefined)
+        }
+
+        render() {
+            const self = this;
```

Comment:
`self` is only being used in the `.map()` call, which ought to have the same `this` anyway once you replace `function(itm)` with `itm =>`. So this can be dropped completely. In general, with ES6 and arrow functions, this `self = this` pattern should be obsolete.